<h1 id="readmeforrsem">README for RSEM</h1>

<p><a href="https://lilab-bcb.github.io/">Bo Li</a> (bli28 at mgh dot harvard dot edu)</p>

<hr />

<h2 id="tableofcontents">Table of Contents</h2>

<ul>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#compilation">Compilation &amp; Installation</a></li>
<li><a href="#usage">Usage</a>

<ul>
<li><a href="#built">Build RSEM references using RefSeq, Ensembl, or GENCODE annotations</a></li>
<li><a href="#untypical">Build RSEM references for untypical organisms</a></li>
</ul></li>
<li><a href="#example-main">Example</a></li>
<li><a href="#simulation">Simulation</a></li>
<li><a href="#gen_trinity">Generate Transcript-to-Gene-Map from Trinity Output</a></li>
<li><a href="#de">Differential Expression Analysis</a></li>
<li><a href="#pRSEM">Prior-Enhanced RSEM (pRSEM)</a></li>
<li><a href="#authors">Authors</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
<li><a href="#license">License</a></li>
</ul>

<hr />

<h2 id="anameintroductionaintroduction"><a name="introduction"></a> Introduction</h2>

<p>RSEM is a software package for estimating gene and isoform expression
levels from RNA-Seq data. The RSEM package provides an user-friendly
interface, supports threads for parallel computation of the EM
algorithm, single-end and paired-end read data, quality scores,
variable-length reads and RSPD estimation. In addition, it provides
posterior mean and 95% credibility interval estimates for expression
levels. For visualization, It can generate BAM and Wiggle files in
both transcript-coordinate and genomic-coordinate. Genomic-coordinate
files can be visualized by both UCSC Genome browser and Broad
Institute&#8217;s Integrative Genomics Viewer (IGV). Transcript-coordinate
files can be visualized by IGV. RSEM also has its own scripts to
generate transcript read depth plots in pdf format. The unique feature
of RSEM is, the read depth plots can be stacked, with read depth
contributed to unique reads shown in black and contributed to
multi-reads shown in red. In addition, models learned from data can
also be visualized. Last but not least, RSEM contains a simulator.</p>

<h2 id="anamecompilationacompilationinstallation"><a name="compilation"></a> Compilation &amp; Installation</h2>

<p>To compile RSEM, simply run</p>

<pre><code>make
</code></pre>

<p>For Cygwin users, run</p>

<pre><code>make cygwin=true
</code></pre>

<p>To compile EBSeq, which is included in the RSEM package, run</p>

<pre><code>make ebseq
</code></pre>

<p>To install RSEM, simply put the RSEM directory in your environment&#8217;s PATH
variable. Alternatively, run</p>

<pre><code>make install
</code></pre>

<p>By default, RSEM executables are installed to <code>/usr/local/bin</code>. You
can change the installation location by setting <code>DESTDIR</code> and/or
<code>prefix</code> variables. The RSEM executables will be installed to
<code>\({DESTDIR}\){prefix}/bin</code>. The default values of <code>DESTDIR</code> and
<code>prefix</code> are <code>DESTDIR=</code> and <code>prefix=/usr/local</code>. For example,</p>

<pre><code>make install DESTDIR=/home/my_name prefix=/software
</code></pre>

<p>will install RSEM executables to <code>/home/my_name/software/bin</code>.</p>

<p><strong>Note</strong> that <code>make install</code> does not install <code>EBSeq</code> related scripts,
such as <code>rsem-generate-ngvector</code>, <code>rsem-run-ebseq</code>, and
<code>rsem-control-fdr</code>. But <code>rsem-generate-data-matrix</code>, which generates
count matrix for differential expression analysis, is installed.</p>

<h3 id="prerequisites">Prerequisites</h3>

<p>C++, Perl and R are required to be installed.</p>

<p>To use the <code>--gff3</code> option of <code>rsem-prepare-reference</code>, Python is also
required to be installed.</p>

<p>To take advantage of RSEM&#8217;s built-in support for the Bowtie/Bowtie
2/STAR alignment program, you must have
<a href="http://bowtie-bio.sourceforge.net">Bowtie</a>/<a href="http://bowtie-bio.sourceforge.net/bowtie2">Bowtie
2</a>/<a href="https://github.com/alexdobin/STAR">STAR</a>
installed.</p>

<h2 id="anameusageausage"><a name="usage"></a> Usage</h2>

<h3 id="i.preparingreferencesequences">I. Preparing Reference Sequences</h3>

<p>RSEM can extract reference transcripts from a genome if you provide it
with gene annotations in a GTF/GFF3 file. Alternatively, you can provide
RSEM with transcript sequences directly.</p>

<p>Please note that GTF files generated from the UCSC Table Browser do not
contain isoform-gene relationship information. However, if you use the
UCSC Genes annotation track, this information can be recovered by
downloading the knownIsoforms.txt file for the appropriate genome.</p>

<p>To prepare the reference sequences, you should run the
<code>rsem-prepare-reference</code> program. Run</p>

<pre><code>rsem-prepare-reference --help
</code></pre>

<p>to get usage information or visit the <a href="rsem-prepare-reference.html">rsem-prepare-reference
documentation page</a>.</p>

<h4 id="anamebuiltabuildrsemreferencesusingrefseqensemblorgencodeannotations"><a name="built"></a> Build RSEM references using RefSeq, Ensembl, or GENCODE annotations</h4>

<p>RefSeq and Ensembl are two frequently used annotations. For human and
mouse, GENCODE annotaions are also available. In this section, we show
how to build RSEM references using these annotations. Note that it is
important to pair the genome with the annotation file for each
annotation source. In addition, we recommend users to use the primary
assemblies of genomes. Without loss of generality, we use human genome as
an example and in addition build Bowtie indices.</p>

<p>For <strong>RefSeq</strong>, the genome and annotation file in GFF3 format can be found
at RefSeq genomes FTP:</p>

<pre><code>ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/
</code></pre>

<p>For example, the human genome and GFF3 file locate at the subdirectory
<code>vertebrate_mammalian/Homo_sapiens/all_assembly_versions/GCF_000001405.31_GRCh38.p5</code>. <code>GCF_000001405.31_GRCh38.p5</code>
is the latest annotation version when this section was written.</p>

<p>Download and decompress the genome and annotation files to your working directory:</p>

<pre><code>ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Homo_sapiens/all_assembly_versions/GCF_000001405.31_GRCh38.p5/GCF_000001405.31_GRCh38.p5_genomic.fna.gz
ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Homo_sapiens/all_assembly_versions/GCF_000001405.31_GRCh38.p5/GCF_000001405.31_GRCh38.p5_genomic.gff.gz
</code></pre>

<p><code>GCF_000001405.31_GRCh38.p5_genomic.fna</code> contains all top level
sequences, including patches and haplotypes. To obtain the primary
assembly, run the following RSEM python script:</p>

<pre><code>rsem-refseq-extract-primary-assembly GCF_000001405.31_GRCh38.p5_genomic.fna GCF_000001405.31_GRCh38.p5_genomic.primary_assembly.fna
</code></pre>

<p>Then type the following command to build RSEM references:</p>

<pre><code>rsem-prepare-reference --gff3 GCF_000001405.31_GRCh38.p5_genomic.gff \
		       --trusted-sources BestRefSeq,Curated\ Genomic \
		       --bowtie \
		       GCF_000001405.31_GRCh38.p5_genomic.primary_assembly.fna \
		       ref/human_refseq
</code></pre>

<p>In the above command, <code>--trusted-sources</code> tells RSEM to only extract
transcripts from RefSeq sources like <code>BestRefSeq</code> or <code>Curated Genomic</code>. By
default, RSEM trust all sources. There is also an
<code>--gff3-RNA-patterns</code> option and its default is <code>mRNA</code>. Setting
<code>--gff3-RNA-patterns mRNA,rRNA</code> will allow RSEM to extract all mRNAs
and rRNAs from the genome. Visit <a href="rsem-prepare-reference.html">here</a>
for more details.</p>

<p>Because the gene and transcript IDs (e.g. gene1000, rna28655)
extracted from RefSeq GFF3 files are hard to understand, it is
recommended to turn on the <code>--append-names</code> option in
<code>rsem-calculate-expression</code> for better interpretation of
quantification results.</p>

<p>For <strong>Ensembl</strong>, the genome and annotation files can be found at
<a href="http://uswest.ensembl.org/info/data/ftp/index.html">Ensembl FTP</a>.</p>

<p>Download and decompress the human genome and GTF files:</p>

<pre><code>ftp://ftp.ensembl.org/pub/release-83/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz
ftp://ftp.ensembl.org/pub/release-83/gtf/homo_sapiens/Homo_sapiens.GRCh38.83.gtf.gz
</code></pre>

<p>Then use the following command to build RSEM references:</p>

<pre><code>rsem-prepare-reference --gtf Homo_sapiens.GRCh38.83.gtf \
		       --bowtie \
		       Homo_sapiens.GRCh38.dna.primary_assembly.fa \
		       ref/human_ensembl
</code></pre>

<p>If you want to use GFF3 file instead, which is unnecessary and not
recommended, you should add option <code>--gff3-RNA-patterns transcript</code>
because <code>mRNA</code> is replaced by <code>transcript</code> in Ensembl GFF3 files.</p>

<p><strong>GENCODE</strong> only provides human and mouse annotations. The genome and
annotation files can be found from <a href="http://www.gencodegenes.org/">GENCODE
website</a>.</p>

<p>Download and decompress the human genome and GTF files:</p>

<pre><code>ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_24/GRCh38.primary_assembly.genome.fa.gz
ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_24/gencode.v24.annotation.gtf.gz
</code></pre>

<p>Then type the following command:</p>

<pre><code>rsem-prepare-reference --gtf gencode.v24.annotation.gtf \
		       --bowtie \
		       GRCh38.primary_assembly.genome.fa \
		       ref/human_gencode
</code></pre>

<p>Similar to Ensembl annotation, if you want to use GFF3 files (not
recommended), add option <code>--gff3-RNA-patterns transcript</code>.</p>

<h4 id="anameuntypicalabuildrsemreferencesforuntypicalorganisms"><a name="untypical"></a> Build RSEM references for untypical organisms</h4>

<p>For untypical organisms, such as viruses, you may only have a GFF3 file that containing only genes but not any transcripts. You need to turn on <code>--gff3-genes-as-transcripts</code> so that RSEM will make each gene as a unique transcript.</p>

<p>Here is an example command:</p>

<pre><code>rsem-prepare-reference --gff3 virus.gff \
               --gff3-genes-as-transcripts \
               --bowtie \
               virus.genome.fa \
               ref/virus
</code></pre>

<h3 id="ii.calculatingexpressionvalues">II. Calculating Expression Values</h3>

<p>To calculate expression values, you should run the
<code>rsem-calculate-expression</code> program. Run</p>

<pre><code>rsem-calculate-expression --help
</code></pre>

<p>to get usage information or visit the <a href="rsem-calculate-expression.html">rsem-calculate-expression
documentation page</a>.</p>

<h4 id="calculatingexpressionvaluesfromsingle-enddata">Calculating expression values from single-end data</h4>

<p>For single-end models, users have the option of providing a fragment
length distribution via the <code>--fragment-length-mean</code> and
<code>--fragment-length-sd</code> options. The specification of an accurate fragment
length distribution is important for the accuracy of expression level
estimates from single-end data. If the fragment length mean and sd are
not provided, RSEM will not take a fragment length distribution into
consideration.</p>

<h4 id="usinganalternativealigner">Using an alternative aligner</h4>

<p>By default, RSEM automates the alignment of reads to reference
transcripts using the Bowtie aligner. Turn on <code>--bowtie2</code> for
<code>rsem-prepare-reference</code> and <code>rsem-calculate-expression</code> will allow
RSEM to use the Bowtie 2 alignment program instead. Please note that
indel alignments, local alignments and discordant alignments are
disallowed when RSEM uses Bowtie 2 since RSEM currently cannot handle
them. See the description of <code>--bowtie2</code> option in
<code>rsem-calculate-expression</code> for more details. Similarly, turn on
<code>--star</code> will allow RSEM to use the STAR aligner. To use an
alternative alignment program, align the input reads against the file
<code>reference_name.idx.fa</code> generated by <code>rsem-prepare-reference</code>, and
format the alignment output in SAM/BAM/CRAM format. Then, instead of
providing reads to <code>rsem-calculate-expression</code>, specify the
<code>--alignments</code> option and provide the SAM/BAM/CRAM file as an
argument.</p>

<p>RSEM requires the alignments of a read to be adjacent. For paired-end
reads, RSEM also requires the two mates of any alignment be
adjacent. To check if your SAM/BAM/CRAM file satisfy the requirements,
run</p>

<pre><code>rsem-sam-validator &lt;input.sam/input.bam/input.cram&gt;
</code></pre>

<p>If your file does not satisfy the requirements, you can use
<code>convert-sam-for-rsem</code> to convert it into a BAM file which RSEM can
process. Run</p>

<pre><code>convert-sam-for-rsem --help
</code></pre>

<p>to get usage information or visit the <a href="convert-sam-for-rsem.html">convert-sam-for-rsem
documentation
page</a>.</p>

<p>Note that RSEM does ** not ** support gapped alignments. So make sure
that your aligner does not produce alignments with
intersions/deletions. In addition, you should make sure that you use
<code>reference_name.idx.fa</code>, which is generated by RSEM, to build your
aligner&#8217;s indices.</p>

<h3 id="iii.visualization">III. Visualization</h3>

<p>RSEM includes a copy of SAMtools. When <code>--no-bam-output</code> is not
specified and <code>--sort-bam-by-coordinate</code> is specified, RSEM will
produce these three files:<code>sample_name.transcript.bam</code>, the unsorted
BAM file, <code>sample_name.transcript.sorted.bam</code> and
<code>sample_name.transcript.sorted.bam.bai</code> the sorted BAM file and
indices generated by the SAMtools included. All three files are in
transcript coordinates. When users in addition specify the
<code>--output-genome-bam</code> option, RSEM will produce three more files:
<code>sample_name.genome.bam</code>, the unsorted BAM file,
<code>sample_name.genome.sorted.bam</code> and
<code>sample_name.genome.sorted.bam.bai</code> the sorted BAM file and
indices. All these files are in genomic coordinates.</p>

<h4 id="aconvertingtranscriptbamfileintogenomebamfile">a) Converting transcript BAM file into genome BAM file</h4>

<p>Normally, RSEM will do this for you via <code>--output-genome-bam</code> option
of <code>rsem-calculate-expression</code>. However, if you have run
<code>rsem-prepare-reference</code> and use <code>reference_name.idx.fa</code> to build
indices for your aligner, you can use <code>rsem-tbam2gbam</code> to convert your
transcript coordinate BAM alignments file into a genomic coordinate
BAM alignments file without the need to run the whole RSEM
pipeline.</p>

<p>Usage:</p>

<pre><code>rsem-tbam2gbam reference_name unsorted_transcript_bam_input genome_bam_output
</code></pre>

<p>reference_name	 		 : The name of reference built by <code>rsem-prepare-reference</code>				
unsorted_transcript_bam_input	 : This file should satisfy: 1) the alignments of a same read are grouped together, 2) for any paired-end alignment, the two mates should be adjacent to each other, 3) this file should not be sorted by samtools
genome_bam_output		 : The output genomic coordinate BAM file&#8217;s name</p>

<h4 id="bgeneratingawigglefile">b) Generating a Wiggle file</h4>

<p>A wiggle plot representing the expected number of reads overlapping
each position in the genome/transcript set can be generated from the
sorted genome/transcript BAM file output. To generate the wiggle
plot, run the <code>rsem-bam2wig</code> program on the
<code>sample_name.genome.sorted.bam</code>/<code>sample_name.transcript.sorted.bam</code> file.</p>

<p>Usage:</p>

<pre><code>rsem-bam2wig sorted_bam_input wig_output wiggle_name [--no-fractional-weight]
</code></pre>

<p>sorted_bam_input  : Input BAM format file, must be sorted<br />
wig_output    : Output wiggle file&#8217;s name, e.g. output.wig<br />
wiggle_name    : The name of this wiggle plot<br />
&#8211;no-fractional-weight : If this is set, RSEM will not look for &#8220;ZW&#8221; tag and each alignment appeared in the BAM file has weight 1. Set this if your BAM file is not generated by RSEM. Please note that this option must be at the end of the command line</p>

<h4 id="cloadingabamandorwigglefileintotheucscgenomebrowserorintegrativegenomicsviewerigv">c) Loading a BAM and/or Wiggle file into the UCSC Genome Browser or Integrative Genomics Viewer(IGV)</h4>

<p>For UCSC genome browser, please refer to the <a href="http://genome.ucsc.edu/goldenPath/help/customTrack.html">UCSC custom track help page</a>.</p>

<p>For integrative genomics viewer, please refer to the <a href="http://www.broadinstitute.org/software/igv/home">IGV home page</a>. Note: Although IGV can generate read depth plot from the BAM file given, it cannot recognize &#8220;ZW&#8221; tag RSEM puts. Therefore IGV counts each alignment as weight 1 instead of the expected weight for the plot it generates. So we recommend to use the wiggle file generated by RSEM for read depth visualization.</p>

<p>Here are some guidance for visualizing transcript coordinate files using IGV:</p>

<p>1) Import the transcript sequences as a genome</p>

<p>Select File -&gt; Import Genome, then fill in ID, Name and Fasta file. Fasta file should be <code>reference_name.idx.fa</code>. After that, click Save button. Suppose ID is filled as <code>reference_name</code>, a file called <code>reference_name.genome</code> will be generated. Next time, we can use: File -&gt; Load Genome, then select <code>reference_name.genome</code>.</p>

<p>2) Load visualization files</p>

<p>Select File -&gt; Load from File, then choose one transcript coordinate visualization file generated by RSEM. IGV might require you to convert wiggle file to tdf file. You should use igvtools to perform this task. One way to perform the conversion is to use the following command:</p>

<pre><code>igvtools tile reference_name.transcript.wig reference_name.transcript.tdf reference_name.genome   
</code></pre>

<h4 id="dgeneratingtranscriptwiggleplots">d) Generating Transcript Wiggle Plots</h4>

<p>To generate transcript wiggle plots, you should run the
<code>rsem-plot-transcript-wiggles</code> program. Run</p>

<pre><code>rsem-plot-transcript-wiggles --help
</code></pre>

<p>to get usage information or visit the <a href="rsem-plot-transcript-wiggles.html">rsem-plot-transcript-wiggles
documentation page</a>.</p>

<h4 id="evisualizethemodellearnedbyrsem">e) Visualize the model learned by RSEM</h4>

<p>RSEM provides an R script, <code>rsem-plot-model</code>, for visulazing the model learned.</p>

<p>Usage:</p>

<pre><code>rsem-plot-model sample_name output_plot_file
</code></pre>

<p>sample_name: the name of the sample analyzed<br />
output_plot_file: the file name for plots generated from the model. It is a pdf file</p>

<p>The plots generated depends on read type and user configuration. It
may include fragment length distribution, mate length distribution,
read start position distribution (RSPD), quality score vs observed
quality given a reference base, position vs percentage of sequencing
error given a reference base and alignment statistics.</p>

<p>fragment length distribution and mate length distribution: x-axis is fragment/mate length, y axis is the probability of generating a fragment/mate with the associated length</p>

<p>RSPD: Read Start Position Distribution. x-axis is bin number, y-axis is the probability of each bin. RSPD can be used as an indicator of 3' bias</p>

<p>Quality score vs. observed quality given a reference base: x-axis is Phred quality scores associated with data, y-axis is the &#8220;observed quality&#8221;, Phred quality scores learned by RSEM from the data. Q = -10log_10(P), where Q is Phred quality score and P is the probability of sequencing error for a particular base</p>

<p>Position vs. percentage sequencing error given a reference base: x-axis is position and y-axis is percentage sequencing error</p>

<p>Alignment statistics: It includes a histogram and a pie chart. For the histogram, x-axis shows the number of <strong>isoform-level</strong> alignments a read has and y-axis provides the number of reads with that many alignments. The inf in x-axis means number of reads filtered due to too many alignments. For the pie chart, four categories of reads &#8212; unalignable, unique, <strong>isoform-level</strong>multi-mapping, filtered &#8211; are plotted and their percentages are noted. In both the histogram and the piechart, numbers belong to unalignable, unique, multi-mapping, and filtered are colored as green, blue, gray and red.</p>

<h2 id="anameexample-mainaexample"><a name="example-main"></a> Example</h2>

<p>Suppose we download the mouse genome from UCSC Genome Browser. We do
not add poly(A) tails and use <code>/ref/mouse_0</code> as the reference name.
We have a FASTQ-formatted file, <code>mmliver.fq</code>, containing single-end
reads from one sample, which we call <code>mmliver_single_quals</code>. We want
to estimate expression values by using the single-end model with a
fragment length distribution. We know that the fragment length
distribution is approximated by a normal distribution with a mean of
150 and a standard deviation of 35. We wish to generate 95%
credibility intervals in addition to maximum likelihood estimates.
RSEM will be allowed 1G of memory for the credibility interval
calculation. We will visualize the probabilistic read mappings
generated by RSEM on UCSC genome browser. We will generate a list of
genes<code>transcript wiggle plots in</code>output.pdf<code>. The list is</code>gene_ids.txt<code>. We will visualize the models learned in</code>mmliver_single_quals.models.pdf`</p>

<p>The commands for this scenario are as follows:</p>

<pre><code>rsem-prepare-reference --gtf mm9.gtf --transcript-to-gene-map knownIsoforms.txt --bowtie --bowtie-path /sw/bowtie /data/mm9 /ref/mouse_0
rsem-calculate-expression --bowtie-path /sw/bowtie --phred64-quals --fragment-length-mean 150.0 --fragment-length-sd 35.0 -p 8 --output-genome-bam --calc-ci --ci-memory 1024 /data/mmliver.fq /ref/mouse_0 mmliver_single_quals
rsem-bam2wig mmliver_single_quals.sorted.bam mmliver_single_quals.sorted.wig mmliver_single_quals
rsem-plot-transcript-wiggles --gene-list --show-unique mmliver_single_quals gene_ids.txt output.pdf 
rsem-plot-model mmliver_single_quals mmliver_single_quals.models.pdf
</code></pre>

<h2 id="anamesimulationasimulation"><a name="simulation"></a> Simulation</h2>

<p>RSEM provides users the <code>rsem-simulate-reads</code> program to simulate RNA-Seq data based on parameters learned from real data sets. Run</p>

<pre><code>rsem-simulate-reads
</code></pre>

<p>to get usage information or read the following subsections.</p>

<h3 id="usage:">Usage:</h3>

<pre><code>rsem-simulate-reads reference_name estimated_model_file estimated_isoform_results theta0 N output_name [-q]
</code></pre>

<p><strong>reference_name:</strong> The name of RSEM references, which should be already generated by <code>rsem-prepare-reference</code> 	</p>

<p><strong>estimated_model_file:</strong> This file describes how the RNA-Seq reads will be sequenced given the expression levels. It determines what kind of reads will be simulated (single-end/paired-end, w/o quality score) and includes parameters for fragment length distribution, read start position distribution, sequencing error models, etc. Normally, this file should be learned from real data using <code>rsem-calculate-expression</code>. The file can be found under the <code>sample_name.stat</code> folder with the name of <code>sample_name.model</code>. <code>model_file_description.txt</code> provides the format and meanings of this file.</p>

<p><strong>estimated_isoform_results:</strong> This file contains expression levels for all isoforms recorded in the reference. It can be learned using <code>rsem-calculate-expression</code> from real data. The corresponding file users want to use is <code>sample_name.isoforms.results</code>. If simulating from user-designed expression profile is desired, start from a learned <code>sample_name.isoforms.results</code> file and only modify the <code>TPM</code> column. The simulator only reads the TPM column. But keeping the file format the same is required. If the RSEM references built are aware of allele-specific transcripts, <code>sample_name.alleles.results</code> should be used instead.</p>

<p><strong>theta0:</strong> This parameter determines the fraction of reads that are coming from background &#8220;noise&#8221; (instead of from a transcript). It can also be estimated using <code>rsem-calculate-expression</code> from real data. Users can find it as the first value of the third line of the file <code>sample_name.stat/sample_name.theta</code>.</p>

<p><strong>N:</strong> The total number of reads to be simulated. If <code>rsem-calculate-expression</code> is executed on a real data set, the total number of reads can be found as the 4th number of the first line of the file <code>sample_name.stat/sample_name.cnt</code>.</p>

<p><strong>output_name:</strong> Prefix for all output files.</p>

<p><strong>&#8211;seed seed:</strong> Set seed for the random number generator used in simulation. The seed should be a 32-bit unsigned integer.</p>

<p><strong>-q:</strong> Set it will stop outputting intermediate information.</p>

<h3 id="outputs:">Outputs:</h3>

<p>output_name.sim.isoforms.results, output_name.sim.genes.results: Expression levels estimated by counting where each simulated read comes from.
output_name.sim.alleles.results: Allele-specific expression levels estimated by counting where each simulated read comes from.</p>

<p>output_name.fa if single-end without quality score;<br />
output_name.fq if single-end with quality score;<br />
output_name_1.fa &amp; output_name_2.fa if paired-end without quality
score;<br />
output_name_1.fq &amp; output_name_2.fq if paired-end with quality score.</p>

<p><strong>Format of the header line</strong>: Each simulated read&#8217;s header line encodes where it comes from. The header line has the format:</p>

<pre><code>{&gt;/@}_rid_dir_sid_pos[_insertL]
</code></pre>

<p><strong>{&gt;/@}:</strong> Either &#8216;&gt;&#8217; or &#8216;@&#8217; must appear. &#8216;&gt;&#8217; appears if FASTA files are generated and &#8216;@&#8217; appears if FASTQ files are generated</p>

<p><strong>rid:</strong> Simulated read&#8217;s index, numbered from 0</p>

<p><strong>dir:</strong> The direction of the simulated read. 0 refers to forward strand (&#8216;+&#8217;) and 1 refers to reverse strand (&#8216;-&#8217;)</p>

<p><strong>sid:</strong> Represent which transcript this read is simulated from. It ranges between 0 and M, where M is the total number of transcripts. If sid=0, the read is simulated from the background noise. Otherwise, the read is simulated from a transcript with index sid. Transcript sid&#8217;s transcript name can be found in the <code>transcript_id</code> column of the <code>sample_name.isoforms.results</code> file (at line sid + 1, line 1 is for column names)</p>

<p><strong>pos:</strong> The start position of the simulated read in strand dir of transcript sid. It is numbered from 0</p>

<p><strong>insertL:</strong> Only appear for paired-end reads. It gives the insert length of the simulated read.</p>

<h3 id="example:">Example:</h3>

<p>Suppose we want to simulate 50 millon single-end reads with quality scores and use the parameters learned from <a href="#example-main">Example</a>. In addition, we set theta0 as 0.2 and output_name as <code>simulated_reads</code>. The command is:</p>

<pre><code>rsem-simulate-reads /ref/mouse_0 mmliver_single_quals.stat/mmliver_single_quals.model mmliver_single_quals.isoforms.results 0.2 50000000 simulated_reads
</code></pre>

<h2 id="anamegen_trinityageneratetranscript-to-gene-mapfromtrinityoutput"><a name="gen_trinity"></a> Generate Transcript-to-Gene-Map from Trinity Output</h2>

<p>For Trinity users, RSEM provides a perl script to generate transcript-to-gene-map file from the fasta file produced by Trinity.</p>

<h3 id="usage:">Usage:</h3>

<pre><code>extract-transcript-to-gene-map-from-trinity trinity_fasta_file map_file
</code></pre>

<p>trinity_fasta_file: the fasta file produced by trinity, which contains all transcripts assembled.<br />
map_file: transcript-to-gene-map file&#8217;s name.</p>

<h2 id="anamedeadifferentialexpressionanalysis"><a name="de"></a> Differential Expression Analysis</h2>

<p>Popular differential expression (DE) analysis tools such as edgeR and
DESeq do not take variance due to read mapping uncertainty into
consideration. Because read mapping ambiguity is prevalent among
isoforms and de novo assembled transcripts, these tools are not ideal
for DE detection in such conditions.</p>

<p>EBSeq, an empirical Bayesian DE analysis tool developed in UW-Madison,
can take variance due to read mapping ambiguity into consideration by
grouping isoforms with parent gene&#8217;s number of isoforms. In addition,
it is more robust to outliers. For more information about EBSeq
(including the paper describing their method), please visit <a href="http://www.biostat.wisc.edu/~ningleng/EBSeq_Package">EBSeq&#8217;s
website</a>.</p>

<p>RSEM includes EBSeq in its folder named <code>EBSeq</code>. To use it, first type</p>

<pre><code>make ebseq
</code></pre>

<p>to compile the EBSeq related codes.</p>

<p>EBSeq requires gene-isoform relationship for its isoform DE
detection. However, for de novo assembled transcriptome, it is hard to
obtain an accurate gene-isoform relationship. Instead, RSEM provides a
script <code>rsem-generate-ngvector</code>, which clusters transcripts based on
measures directly relating to read mappaing ambiguity. First, it
calculates the &#8216;unmappability&#8217; of each transcript. The &#8216;unmappability&#8217;
of a transcript is the ratio between the number of k mers with at
least one perfect match to other transcripts and the total number of k
mers of this transcript, where k is a parameter. Then, Ng vector is
generated by applying Kmeans algorithm to the &#8216;unmappability&#8217; values
with number of clusters set as 3. This program will make sure the mean
&#8216;unmappability&#8217; scores for clusters are in ascending order. All
transcripts whose lengths are less than k are assigned to cluster</p>

<ol>
<li>Run

rsem-generate-ngvector &#8211;help</li>
</ol>

<p>to get usage information or visit the <a href="rsem-generate-ngvector.html">rsem-generate-ngvector
documentation
page</a>.</p>

<p>If your reference is a de novo assembled transcript set, you should
run <code>rsem-generate-ngvector</code> first. Then load the resulting
<code>output_name.ngvec</code> into R. For example, you can use</p>

<pre><code>NgVec &lt;- scan(file=&quot;output_name.ngvec&quot;, what=0, sep=&quot;\n&quot;)
</code></pre>

<p>. After that, set &#8220;NgVector = NgVec&#8221; for your differential expression
test (either <code>EBTest</code> or <code>EBMultiTest</code>).</p>

<p>For users' convenience, RSEM also provides a script
<code>rsem-generate-data-matrix</code> to extract input matrix from expression
results:</p>

<pre><code>rsem-generate-data-matrix sampleA.[genes/isoforms].results sampleB.[genes/isoforms].results ... &gt; output_name.counts.matrix
</code></pre>

<p>The results files are required to be either all gene level results or
all isoform level results. You can load the matrix into R by</p>

<pre><code>IsoMat &lt;- data.matrix(read.table(file=&quot;output_name.counts.matrix&quot;))
</code></pre>

<p>before running either <code>EBTest</code> or <code>EBMultiTest</code>.</p>

<p>Lastly, RSEM provides two scripts, <code>rsem-run-ebseq</code> and
<code>rsem-control-fdr</code>, to help users find differential expressed
genes/transcripts. First, <code>rsem-run-ebseq</code> calls EBSeq to calculate related statistics
for all genes/transcripts. Run</p>

<pre><code>rsem-run-ebseq --help
</code></pre>

<p>to get usage information or visit the <a href="rsem-run-ebseq.html">rsem-run-ebseq documentation
page</a>. Second,
<code>rsem-control-fdr</code> takes <code>rsem-run-ebseq</code> 's result and reports called
differentially expressed genes/transcripts by controlling the false
discovery rate. Run</p>

<pre><code>rsem-control-fdr --help
</code></pre>

<p>to get usage information or visit the <a href="rsem-control-fdr.html">rsem-control-fdr documentation
page</a>. These
two scripts can perform DE analysis on either 2 conditions or multiple
conditions.</p>

<p>Please note that <code>rsem-run-ebseq</code> and <code>rsem-control-fdr</code> use EBSeq&#8217;s
default parameters. For advanced use of EBSeq or information about how
EBSeq works, please refer to <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/EBSeq/inst/doc/EBSeq_Vignette.pdf">EBSeq&#8217;s
manual</a>.</p>

<p>Questions related to EBSeq should
be sent to <a href="mailto:nleng@wisc.edu">Ning Leng</a>.</p>

<h2 id="anameprsemaprior-enhancedrsemprsem"><a name="pRSEM"></a> Prior-Enhanced RSEM (pRSEM)</h2>

<h3 id="i.overview">I. Overview</h3>

<p><a href="https://deweylab.github.io/pRSEM/">Prior-enhanced RSEM (pRSEM)</a> uses complementary information (e.g. ChIP-seq data) to allocate RNA-seq multi-mapping fragments. We included pRSEM code in the subfolder <code>pRSEM/</code> as well as in RSEM&#8217;s scripts <code>rsem-prepare-reference</code> and <code>rsem-calculate-expression</code>.</p>

<h3 id="ii.demo">II. Demo</h3>

<p>To get a quick idea on how to use pRSEM, you can try <a href="https://github.com/pliu55/pRSEM_demo">this demo</a>. It provides a single script, named <code>run_pRSEM_demo.sh</code>, which allows you to run all pRSEM&#8217;s functions. It also contains detailed descriptions of pRSEM&#8217;s workflow, input and output files.</p>

<h3 id="iii.installation">III. Installation</h3>

<p>To compile pRSEM, type</p>

<pre><code>make pRSEM
</code></pre>

<p>Note that you need to first compile RSEM before compiling pRSEM. Currently, pRSEM has only been tested on Linux.</p>

<h3 id="iv.example">IV. Example</h3>

<p>To run pRSEM on the <a href="#example-main">RSEM example above</a>, you need to provide:</p>

<ul>
<li><strong>ChIP-seq sequencing file(s) in FASTQ format</strong> or <strong>a ChIP-seq peak file in BED format</strong>. They will be used by pRSEM to obtain complementatry information for allocating RNA-seq multi-mapping fragments.</li>
<li><strong>a genome mappability file in bigWig format</strong> to let pRSEM build a training
set of isoforms to learn prior. Mappability can be obtained from UCSC&#8217;s
ENCODE composite track for <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeCrgMapabilityAlign36mer.bigWig">human hg19</a>
and <a href="http://hgdownload.cse.ucsc.edu/goldenPath/mm9/encodeDCC/wgEncodeMapability/wgEncodeCrgMapabilityAlign36mer.bigWig">mouse mm9</a>. For other genomes, you
can generate the mappability file by following [this tutorial] (http://wiki.bits.vib.be/index.php/Create_a_mappability_track#Install_and_run_the_GEM_library_tools).</li>
</ul>

<p>Assuming you would like to use RNA Pol II&#8217;s ChIP-seq sequencing files <code>/data/mmliver_PolIIRep1.fq.gz</code> and <code>/data/mmliver_PolIIRep2.fq.gz</code>, with ChIP-seq control <code>/data/mmliver_ChIPseqCtrl.fq.gz</code>. Also, assuming the mappability file for mouse genome is <code>/data/mm9.bigWig</code> and you prefer to use STAR located at <code>/sw/STAR</code> to align RNA-seq fragments and use Bowtie to align ChIP-seq reads. Then, you can use the following commands to run pRSEM:</p>

<pre><code>rsem-prepare-reference --gtf mm9.gtf \
                       --star \
                       --star-path /sw/STAR \
                       -p 8 \
                       --prep-pRSEM \
                       --bowtie-path /sw/bowtie \
                       --mappability-bigwig-file /data/mm9.bigWig \
                       /data/mm9 \
                       /ref/mouse_0
  
rsem-calculate-expression --star \
                          --star-path /sw/STAR \
                          --calc-pme \
                          --run-pRSEM \
                          --chipseq-target-read-files /data/mmliver_PolIIRep1.fq.gz,/data/mmliver_PolIIRep2.fq.gz \
                          --chipseq-control-read-files /data/mmliver_ChIPseqCtrl.fq.gz \
                          --bowtie-path /sw/bowtie \
                          -p 8 \
                          /data/mmliver.fq \
                          /ref/mouse_0 \
                          mmliver_single_quals
</code></pre>

<p>To find out more about pRSEM options and examples, you can use the commands:</p>

<pre><code>rsem-prepare-reference --help
</code></pre>

<p>and</p>

<pre><code>rsem-calculate-expression --help
</code></pre>

<h3 id="v.systemrequirements">V. System Requirements</h3>

<ul>
<li>Linux</li>
<li>Perl version &gt;= 5.8.8</li>
<li>Python version &gt;= 2.7.3</li>
<li>R version &gt;= 3.3.1</li>
<li>Bioconductor 3.3</li>
</ul>

<h3 id="vi.requiredexternalpackages">VI. Required External Packages</h3>

<p>All the following packages will be automatically installed when compiling pRSEM.</p>

<ul>
<li><a href="https://cran.r-project.org/web/packages/data.table/index.html">data.table 1.9.6</a>: an extension of R&#8217;s data.frame, heavily used by pRSEM.</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges 1.24.3</a>: efficient representing and manipulating genomic intervals, heavily used by pRSEM.</li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/ShortRead.html">ShortRead 1.30.0</a>: guessing the encoding of ChIP-seq FASTQ file&#8217;s quality score.</li>
<li><a href="https://cran.r-project.org/web/packages/caTools/index.html">caTools 1.17.1</a>: used for SPP Peak Caller.</li>
<li><a href="https://code.google.com/archive/p/phantompeakqualtools/">SPP Peak Caller</a>:
ChIP-seq peak caller. Source code was slightly modified in terms of included headers in order to be compiled under R v3.3.1.</li>
<li><a href="https://sites.google.com/site/anshulkundaje/projects/idr/idrCode.tar.gz?attredirects=0">IDR</a>:
calculating Irreproducible Discovery Rate to call peaks from multiple ChIP-seq replicates.</li>
</ul>

<h2 id="anameauthorsaauthors"><a name="authors"></a> Authors</h2>

<p><a href="http://bli25ucb.github.io/">Bo Li</a> and <a href="https://www.biostat.wisc.edu/~cdewey/">Colin Dewey</a> designed the RSEM algorithm. <a href="http://bli25ucb.github.io/">Bo Li</a> implemented the RSEM software. <a href="https://www.biostat.wisc.edu/~cdewey/group.html">Peng Liu</a> contributed the STAR aligner options and prior-enhanced RSEM (pRSEM).</p>

<h2 id="anameacknowledgementsaacknowledgements"><a name="acknowledgements"></a> Acknowledgements</h2>

<p>RSEM uses the <a href="http://www.boost.org/">Boost C++</a> and
<a href="http://www.htslib.org/">SAMtools</a> libraries. RSEM includes
<a href="http://www.biostat.wisc.edu/~ningleng/EBSeq_Package/">EBSeq</a> for
differential expression analysis.</p>

<p>We thank earonesty, Dr. Samuel Arvidsson, John Marshall, and Michael
R. Crusoe for contributing patches.</p>

<p>We thank Han Lin, j.miller, Jo&euml;l Fillon, Dr. Samuel G. Younkin,
Malcolm Cook, Christina Wells, Uro&#353; &#352;ipeti&#263;,
outpaddling, rekado, and Josh Richer for suggesting possible fixes.</p>

<p><strong>Note</strong> that <code>bam_sort.c</code> of SAMtools is slightly modified so that
<code>samtools sort -n</code> will not move the two mates of paired-end
alignments apart. In addition, we turn on the <code>--without-curses</code>
option when configuring SAMtools and thus SAMtools' curses-based
<code>tview</code> subcommand is not built.</p>

<h2 id="anamelicensealicense"><a name="license"></a> License</h2>

<p>RSEM is licensed under the <a href="http://www.gnu.org/licenses/gpl-3.0.html">GNU General Public License
v3</a>.</p>
