<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>rsem-run-ebseq</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:root@osx004.apple.com" />
</head>

<body style="background-color: white">



<ul id="index">
  <li><a href="#NAME">NAME</a></li>
  <li><a href="#SYNOPSIS">SYNOPSIS</a></li>
  <li><a href="#ARGUMENTS">ARGUMENTS</a></li>
  <li><a href="#OPTIONS">OPTIONS</a></li>
  <li><a href="#DESCRIPTION">DESCRIPTION</a></li>
  <li><a href="#OUTPUT">OUTPUT</a></li>
  <li><a href="#EXAMPLES">EXAMPLES</a></li>
</ul>

<h1 id="NAME">NAME</h1>

<p>rsem-run-ebseq</p>

<h1 id="SYNOPSIS">SYNOPSIS</h1>

<p>rsem-run-ebseq [options] data_matrix_file conditions output_file</p>

<h1 id="ARGUMENTS">ARGUMENTS</h1>

<dl>

<dt id="data_matrix_file"><b>data_matrix_file</b></dt>
<dd>

<p>This file is a m by n matrix. m is the number of genes/transcripts and n is the number of total samples. Each element in the matrix represents the expected count for a particular gene/transcript in a particular sample. Users can use &#39;rsem-generate-data-matrix&#39; to generate this file from expression result files.</p>

</dd>
<dt id="conditions"><b>conditions</b></dt>
<dd>

<p>Comma-separated list of values representing the number of replicates for each condition. For example, &quot;3,3&quot; means the data set contains 2 conditions and each condition has 3 replicates. &quot;2,3,3&quot; means the data set contains 3 conditions, with 2, 3, and 3 replicates for each condition respectively.</p>

</dd>
<dt id="output_file"><b>output_file</b></dt>
<dd>

<p>Output file name.</p>

</dd>
</dl>

<h1 id="OPTIONS">OPTIONS</h1>

<dl>

<dt id="ngvector-file"><b>--ngvector</b> &lt;file&gt;</dt>
<dd>

<p>This option provides the grouping information required by EBSeq for isoform-level differential expression analysis. The file can be generated by &#39;rsem-generate-ngvector&#39;. Turning this option on is highly recommended for isoform-level differential expression analysis. (Default: off)</p>

</dd>
<dt id="h---help"><b>-h/--help</b></dt>
<dd>

<p>Show help information.</p>

</dd>
</dl>

<h1 id="DESCRIPTION">DESCRIPTION</h1>

<p>This program is a wrapper over EBSeq. It performs differential expression analysis and can work on two or more conditions. All genes/transcripts and their associated statistcs are reported in one output file. This program does not control false discovery rate and call differential expressed genes/transcripts. Please use &#39;rsem-control-fdr&#39; to control false discovery rate after this program is finished.</p>

<h1 id="OUTPUT">OUTPUT</h1>

<dl>

<dt id="output_file1"><b>output_file</b></dt>
<dd>

<p>This file reports the calculated statistics for all genes/transcripts. It is written as a matrix with row and column names. The row names are the genes&#39;/transcripts&#39; names. The column names are for the reported statistics.</p>

<p>If there are only 2 different conditions among the samples, four statistics (columns) will be reported for each gene/transcript. They are &quot;PPEE&quot;, &quot;PPDE&quot;, &quot;PostFC&quot; and &quot;RealFC&quot;. &quot;PPEE&quot; is the posterior probability (estimated by EBSeq) that a gene/transcript is equally expressed. &quot;PPDE&quot; is the posterior probability that a gene/transcript is differentially expressed. &quot;PostFC&quot; is the posterior fold change (condition 1 over condition2) for a gene/transcript. It is defined as the ratio between posterior mean expression estimates of the gene/transcript for each condition. &quot;RealFC&quot; is the real fold change (condition 1 over condition2) for a gene/transcript. It is the ratio of the normalized within condition 1 mean count over normalized within condition 2 mean count for the gene/transcript. Fold changes are calculated using EBSeq&#39;s &#39;PostFC&#39; function. The genes/transcripts are reported in descending order of their &quot;PPDE&quot; values.</p>

<p>If there are more than 2 different conditions among the samples, the output format is different. For differential expression analysis with more than 2 conditions, EBSeq will enumerate all possible expression patterns (on which conditions are equally expressed and which conditions are not). Suppose there are k different patterns, the first k columns of the output file give the posterior probability of each expression pattern is true. Patterns are defined in a separate file, &#39;output_file.pattern&#39;. The k+1 column gives the maximum a posteriori (MAP) expression pattern for each gene/transcript. The k+2 column gives the posterior probability that not all conditions are equally expressed (column name &quot;PPDE&quot;). The genes/transcripts are reported in descending order of their &quot;PPDE&quot; column values. For details on how EBSeq works for more than 2 conditions, please refer to EBSeq&#39;s manual.</p>

</dd>
<dt id="output_file.pattern"><b>output_file.pattern</b></dt>
<dd>

<p>This file is only generated when there are more than 2 conditions. It defines all possible expression patterns over the conditions using a matrix with names. Each row of the matrix refers to a different expression pattern and each column gives the expression status of a different condition. Two conditions are equally expressed if and only if their statuses are the same.</p>

</dd>
<dt id="output_file.condmeans"><b>output_file.condmeans</b></dt>
<dd>

<p>This file is only generated when there are more than 2 conditions. It gives the normalized mean count value for each gene/transcript at each condition. It is formatted as a matrix with names. Each row represents a gene/transcript and each column represent a condition. The order of genes/transcripts is the same as &#39;output_file&#39;. This file can be used to calculate fold changes between conditions which users are interested in.</p>

</dd>
</dl>

<h1 id="EXAMPLES">EXAMPLES</h1>

<p>1) We&#39;re interested in isoform-level differential expression analysis and there are two conditions. Each condition has 5 replicates. We have already collected the data matrix as &#39;IsoMat&#39; and generated ngvector as &#39;ngvector.ngvec&#39;:</p>

<pre><code> rsem-run-ebseq --ngvector ngvector.ngvec IsoMat 5,5 IsoMat.results</code></pre>

<p>The results will be in &#39;IsoMat.results&#39;.</p>

<p>2) We&#39;re interested in gene-level analysis and there are 3 conditions. The first condition has 3 replicates and the other two has 4 replicates each. The data matrix is named as &#39;GeneMat&#39;:</p>

<pre><code> rsem-run-ebseq GeneMat 3,4,4 GeneMat.results</code></pre>

<p>Three files, &#39;GeneMat.results&#39;, &#39;GeneMat.results.pattern&#39;, and &#39;GeneMat.results.condmeans&#39;, will be generated.</p>


</body>

</html>


